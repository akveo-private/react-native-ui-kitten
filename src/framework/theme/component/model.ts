// TODO(@type): declare theme types

export type ThemeShape = any;
export type ThemedStyleShape = any;
export type StyleSheetShape = any;
