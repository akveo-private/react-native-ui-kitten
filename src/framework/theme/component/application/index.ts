export {
  ApplicationProvider,
  Props as ApplicationProviderProps,
} from './applicationProvider.component';
