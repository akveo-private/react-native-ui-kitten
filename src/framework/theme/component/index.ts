export { ThemeProvider } from './themeProvider.component';
export { withTheme } from './themeConsumer.component';
export { withThemedStyles } from './themeConsumerStyled.component';
export { ThemeType } from './type';
