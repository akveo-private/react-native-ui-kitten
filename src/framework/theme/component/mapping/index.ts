export * from './type';
export {
  MappingProvider as ThemeMappingProvider,
  Props as ThemeMappingProviderProps,
} from './mappingProvider.component';
