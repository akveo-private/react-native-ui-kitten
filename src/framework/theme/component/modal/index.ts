export * from './modal.component';
export * from './modalPanel.component';
