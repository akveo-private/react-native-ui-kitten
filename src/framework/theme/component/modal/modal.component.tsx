import React from 'react';
import {
  View,
  ViewProps,
  TouchableWithoutFeedbackProps,
  StyleSheet,
  TouchableWithoutFeedback,
} from 'react-native';

interface ModalComponentProps {
  visible: boolean;
  component: React.ReactElement<any>;
  isBackDropAllowed: boolean;
  identifier: string;
  onCloseModal: (index: string) => void;
}

interface ModalComponentState {
  isVisible: boolean;
}

export class ModalComponent extends React.Component<ModalComponentProps, ModalComponentState> {

  static defaultProps: Partial<ModalComponentProps> = {
    visible: false,
    isBackDropAllowed: false,
  };

  state: ModalComponentState = {
    isVisible: false,
  };

  constructor(props: ModalComponentProps) {
    super(props);
    this.state = { isVisible: props.visible };
  }

  private closeModal: () => void = (): void => {
    this.props.onCloseModal && this.props.onCloseModal(this.props.identifier);
  };

  private closeOnBackDrop: () => void = () => {
    if (this.props.isBackDropAllowed) {
      this.closeModal();
    }
  };

  private renderChild(): React.ReactElement<ViewProps> {
    const { component } = this.props;
    const componentWithCloseProps = React.Children
      .map(component, (child: React.ReactElement<any>) =>
        React.cloneElement(child, {
            onCloseModal: this.closeModal,
          },
        ));
    return (
      <View style={styles.componentContainer}>
        {componentWithCloseProps}
      </View>
    );
  }

  private renderComponent(): React.ReactElement<ViewProps | TouchableWithoutFeedbackProps> {
    return this.props.isBackDropAllowed ?
      (
        <TouchableWithoutFeedback onPress={this.closeOnBackDrop}>
          {this.renderChild()}
        </TouchableWithoutFeedback>
      ) : this.renderChild();
  }

  render(): React.ReactElement<ViewProps | TouchableWithoutFeedbackProps> {
    return this.state.isVisible ? this.renderComponent() : null;
  }

}

const styles = StyleSheet.create({
  componentContainer: {
    position: 'absolute',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
