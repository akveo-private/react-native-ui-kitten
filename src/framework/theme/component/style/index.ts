export {
  StyleProvider,
  Props as StyleProviderProps,
} from './styleProvider.component';

export {
  styled,
  Props as StyledComponentProps,
} from './styleConsumer.component';
