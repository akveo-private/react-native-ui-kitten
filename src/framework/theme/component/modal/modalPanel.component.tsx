import React from 'react';
import {
  View,
  StyleSheet,
  ViewProps,
} from 'react-native';
import { ModalService } from '../../service';
import { Modal } from '../../../ui/modal/modal.component';

export interface ModalPanelProps {
  children: React.ReactElement<any> | React.ReactElement<any>[];
}

interface ModalPanelState {
  dialogComponents: Map<string, React.ReactElement<any>>;
  backdropValues: Map<string, boolean>;
}

export class ModalPanel extends React.Component<ModalPanelProps, ModalPanelState> {

  public state: ModalPanelState = {
    dialogComponents: new Map(),
    backdropValues: new Map(),
  };

  public componentDidMount(): void {
    ModalService.setComponent(this);
  }

  public componentWillUnmount(): void {
    ModalService.setComponent(null);
  }

  public onCloseModal = (identifier: string) => {
    const components: Map<string, React.ReactElement<any>> = this.state.dialogComponents;
    components.delete(identifier);
    const backdropValues: Map<string, boolean> = this.state.backdropValues;
    backdropValues.delete(identifier);
    this.setState({
      dialogComponents: components,
      backdropValues: backdropValues,
    });
  };

  public showDialog(dialogComponent: React.ReactElement<any>, closeOnBackDrop: boolean): void {
    const key: string = this.generateUniqueComponentKey();
    const componentsMap: Map<string, React.ReactElement<any>> = this.state.dialogComponents
      .set(key, dialogComponent);
    const backdropsMap: Map<string, boolean> = this.state.backdropValues.set(key, closeOnBackDrop);
    this.setState({
      dialogComponents: componentsMap,
      backdropValues: backdropsMap,
    });
  }

  private generateUniqueComponentKey = (): string => {
    return Math.random().toString(36).substring(2);
  };

  private areThereAnyComponents(): boolean {
    return this.state.dialogComponents && this.state.dialogComponents.size !== 0;
  }

  private renderModal(modal: React.ReactElement<any>, index: number) {
    const allModalKeys: string[] = Array.from(this.state.dialogComponents.keys());
    const identifier: string = allModalKeys
      .find(item => this.state.dialogComponents.get(item) === modal);
    const closeOnBackdrop: boolean = this.state.backdropValues.get(identifier);
    return (
      <Modal
        visible={true}
        isBackDropAllowed={closeOnBackdrop}
        key={index}
        identifier={identifier}
        onCloseModal={this.onCloseModal}
      >
        {modal}
      </Modal>
    );
  }

  private renderModals() {
    return Array.from(this.state.dialogComponents.values())
      .map((component: React.ReactElement<any>, i: number) =>
        this.renderModal(component, i));
  }

  public render(): React.ReactElement<ViewProps> {
    return (
      <View style={styles.container}>
        {this.props.children}
        {this.areThereAnyComponents() && this.renderModals()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
