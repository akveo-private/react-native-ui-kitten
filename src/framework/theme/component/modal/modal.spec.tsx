import React from 'react';
import {
  fireEvent,
  render,
} from 'react-native-testing-library';
import {
  View,
  Text,
  TouchableWithoutFeedback,
  Button,
} from 'react-native';
import { ModalPanel } from './modalPanel.component';
import { ModalComponent } from './modal.component';

describe('@modal component/panel checks', () => {

  const MODAL_TEST_IDENTIFIER: (substring: string) => string = (substring: string): string =>
    `modal-test-identifier-${substring}`;

  interface HooksProps {
    componentDidMount?: () => void;
    componentWillUnmount?: () => void;
  }

  class ModalPanelTest extends React.Component<HooksProps> {

    componentDidMount(): void {
      this.props.componentDidMount && this.props.componentDidMount();
    }

    componentWillUnmount() {
      this.props.componentWillUnmount && this.props.componentWillUnmount();
    }

    render() {
      return (
        <ModalPanel>
          <View>
            <Text>Modal Panel Test</Text>
          </View>
        </ModalPanel>
      );
    }
  }

  it('* modal panel renders properly', () => {
    const panel = render(<ModalPanelTest/>);
    expect(panel).toMatchSnapshot();
  });

  it('* modal panel render with props / children checking', () => {
    const panelChild = <View><Text>Test</Text></View>;
    const panel = <ModalPanelTest children={panelChild}/>;
    const renderedPanel = render(panel);

    expect(panel.props.children).toBe(panelChild);
    expect(renderedPanel).toMatchSnapshot();
  });

  it('* modal panel l/c-hooks checks', () => {
    const componentDidMount = jest.fn();
    const componentWillUnmount = jest.fn();

    const wrapper = render(
      <ModalPanelTest
        componentDidMount={componentDidMount}
        componentWillUnmount={componentWillUnmount}
      />,
    );
    expect(componentDidMount).toHaveBeenCalled();
    wrapper.unmount();
    expect(componentWillUnmount).toHaveBeenCalled();
  });

  it('* modal component renders properly', () => {
    const modal1 = render(<ModalComponent
      visible={true}
      component={<View><Text>Test1</Text></View>}
      isBackDropAllowed={false}
      identifier={MODAL_TEST_IDENTIFIER('1')}
      onCloseModal={() => 1}/>);
    const modal2 = render(<ModalComponent
      visible={false}
      component={<View><Text>Test2</Text></View>}
      isBackDropAllowed={false}
      identifier={MODAL_TEST_IDENTIFIER('1')}
      onCloseModal={() => 1}/>);

    expect(modal1).toMatchSnapshot();
    expect(modal2).toMatchSnapshot();
  });

  it('* modal component props checks', () => {
    const modalPassingProps = {
      visible: true,
      component: <View><Text>Test1</Text></View>,
      isBackDropAllowed: false,
    };
    const modal = <ModalComponent {...modalPassingProps}/>;

    expect(modal.props.visible).toBe(modalPassingProps.visible);
    expect(modal.props.component).toBe(modalPassingProps.component);
    expect(modal.props.isBackDropAllowed).toBe(modalPassingProps.isBackDropAllowed);
  });

  it('* modal closes on passed prop', () => {
    const onCloseModal = jest.fn();
    const component = <ModalComponent
      visible={true}
      component={<View>
        <Text>Test1</Text>
        <Button title={'Close Modal'} onPress={onCloseModal}/>
      </View>}
      isBackDropAllowed={true}
      identifier={MODAL_TEST_IDENTIFIER('1')}
      onCloseModal={onCloseModal}/>;
    const modal = render(component);
    expect(modal).toMatchSnapshot();
    fireEvent.press(modal.getByType(Button));
    expect(onCloseModal).toHaveBeenCalled();
    expect(modal).toMatchSnapshot();
  });

  it('* modal component close on backDrop checks', () => {
    const onCloseModal = jest.fn();
    const modal = render(<ModalComponent
      visible={true}
      component={<View><Text>Test1</Text></View>}
      isBackDropAllowed={true}
      identifier={MODAL_TEST_IDENTIFIER('1')}
      onCloseModal={onCloseModal}/>);
    expect(modal).toMatchSnapshot();
    fireEvent.press(modal.getByType(TouchableWithoutFeedback));
    expect(modal).toMatchSnapshot();
    expect(onCloseModal).toHaveBeenCalled();
  });

});
