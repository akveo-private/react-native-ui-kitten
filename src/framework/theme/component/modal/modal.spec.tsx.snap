// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@modal component/panel checks * modal closes on passed prop 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "position": "absolute",
    }
  }
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Close Modal
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`@modal component/panel checks * modal closes on passed prop 2`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "position": "absolute",
    }
  }
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Close Modal
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`@modal component/panel checks * modal component close on backDrop checks 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "position": "absolute",
    }
  }
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
  </View>
</View>
`;

exports[`@modal component/panel checks * modal component close on backDrop checks 2`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "position": "absolute",
    }
  }
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
  </View>
</View>
`;

exports[`@modal component/panel checks * modal component renders properly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "position": "absolute",
    }
  }
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
  </View>
</View>
`;

exports[`@modal component/panel checks * modal component renders properly 2`] = `null`;

exports[`@modal component/panel checks * modal panel render with props / children checking 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <Text>
      Modal Panel Test
    </Text>
  </View>
</View>
`;

exports[`@modal component/panel checks * modal panel renders properly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <Text>
      Modal Panel Test
    </Text>
  </View>
</View>
`;
